core/src/saros/git/GitManager.java

```diff
@@ -0,0 +1,121 @@
+package saros.git;
+
+import java.io.File;
+import java.io.IOException;
+import org.apache.log4j.Logger;
+import org.picocontainer.Startable;
+import saros.activities.GitCollectActivity;
+import saros.activities.GitRequestActivity;
+import saros.activities.GitSendBundleActivity;
+import saros.annotations.Component;
+import saros.session.AbstractActivityConsumer;
+import saros.session.AbstractActivityProducer;
+import saros.session.IActivityConsumer;
+import saros.session.IActivityConsumer.Priority;
+import saros.session.internal.SarosSession;
+
+/**
+ * This Class is responsible for the negotiation between user that want to sent,receive and fetch
+ * commits.
+ *
+ * <p>In order to start the negotiation the user needs to call the procedure {@code
+ * sendCommitRequest()}.
+ *
+ * <p>Before starting negotiation the user need to set up his working directory tree by calling the
+ * procedure {@code changeWorkDirTree(File workDirTree)}. After calling it the first time the
+ * Manager hold a {@link JGitFacade} object.
+ */
+@Component(module = "core")
+public class GitManager extends AbstractActivityProducer implements Startable {
+
+  private static final Logger log = Logger.getLogger(GitManager.class.getName());
+
+  private final SarosSession session;
+
+  JGitFacade jGitFacade;
+
+  public GitManager(SarosSession session) {
+    this.session = session;
+  }
+
+  @Override
+  public void start() {
+    session.addActivityProducer(this);
+    session.addActivityConsumer(consumer, Priority.ACTIVE);
+  }
+
+  @Override
+  public void stop() {
+    session.removeActivityProducer(this);
+    session.removeActivityConsumer(consumer);
+  }
+
+  public void sendCommitRequest() {
+    fireActivity(new GitRequestActivity(session.getLocalUser()));
+  }
+
+  public void changeWorkDirTree(File workDirTree) throws IOException {
+    log.info("You changed your working directory tree");
+    if (jGitFacade == null) {
+      this.jGitFacade = new JGitFacade(workDirTree);
+    } else {
+      this.jGitFacade.setWorkDirTree(workDirTree);
+    }
+  }
+
+  private final IActivityConsumer consumer =
+      new AbstractActivityConsumer() {
+        @Override
+        public void receive(GitRequestActivity activity) {
+          if (jGitFacade != null) {
+            log.info("You recieved a request to fetch commits");
+            try {
+
+              String basis = jGitFacade.getSHA1HashByRevisionString("HEAD");
+              fireActivity(new GitCollectActivity(session.getLocalUser(), basis));
+
+            } catch (IOException e) {
+              log.debug(session.getLocalUser().toString() + " can't access the Git Directory");
```

Comment:
toString() is always invoked when concatenating stings while using objects

Reply:
Are you sure to swallow the stack trace of the exception and log this a debug ?

Reply:
done